crates/vsock/src/main.rs

```diff
@@ -10,38 +10,81 @@ mod vsock_conn;
 
 use std::{convert::TryFrom, sync::Arc};
 
-use clap::Parser;
+use crate::vhu_vsock::{Error, Result, VhostUserVsockBackend, VsockConfig};
+use clap::{Args, Parser};
 use log::{info, warn};
+use serde::Deserialize;
 use vhost::{vhost_user, vhost_user::Listener};
 use vhost_user_backend::VhostUserDaemon;
 use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
 
-use crate::vhu_vsock::{Error, Result, VhostUserVsockBackend, VsockConfig};
-
-#[derive(Parser, Debug)]
-#[clap(version, about, long_about = None)]
-struct VsockArgs {
+#[derive(Args, Debug, Deserialize)]
+struct VsockParam {
     /// Context identifier of the guest which uniquely identifies the device for its lifetime.
-    #[clap(long, default_value_t = 3)]
+    #[arg(long, default_value_t = 3, conflicts_with = "config")]
     guest_cid: u64,
 
     /// Unix socket to which a hypervisor connects to and sets up the control path with the device.
-    #[clap(long)]
+    #[arg(long, conflicts_with = "config")]
     socket: String,
 
     /// Unix socket to which a host-side application connects to.
-    #[clap(long)]
+    #[arg(long, conflicts_with = "config")]
     uds_path: String,
 }
 
+#[derive(Parser, Debug)]
+#[command(version, about, long_about = None)]
+struct VsockArgs {
+    #[command(flatten)]
+    param: Option<VsockParam>,
+
+    /// Load from a given configuration file
+    #[arg(long)]
+    config: Option<String>,
+}
+
+impl VsockArgs {
+    pub fn parse_config(&self) -> Option<VsockConfig> {
+        match &self.config {
+            Some(c) => {
+                match config::Config::builder()
+                    .add_source(config::File::new(c.as_str(), config::FileFormat::Yaml))
+                    .build()
+                {
+                    Ok(s) => {
+                        let mut v = s.get::<Vec<VsockParam>>("vms").unwrap();
+                        if v.len() != 1 {
+                            return None;
+                        }
+                        v.pop()
+                            .map(|vm| VsockConfig::new(vm.guest_cid, vm.socket, vm.uds_path))
+                    }
+                    Err(_) => None,
+                }
+            }
+            _ => None,
+        }
```

Comment:
What about using "if let" for both the match blocks ?

Reply:
```
impl VsockArgs {
    pub fn parse_config(&self) -> Option<VsockConfig> {
        if let Some(s) = &self.config {
            let b = config::Config::builder()
                .add_source(config::File::new(s.as_str(), config::FileFormat::Yaml))
                .build();
            if let Ok(c) = b {
                let mut v = c.get::<Vec<VsockParam>>("vms").unwrap();
                if v.len() != 1 {
                    return None;
                }
                v.pop()
                    .map(|vm| VsockConfig::new(vm.guest_cid, vm.socket, vm.uds_path))
            } else {
                None
            }
        } else {
            None
        }
    }
}
```

You mean something like this?

Reply:
pub fn parse_config(&self) -> Option<VsockConfig> {
        if let Some(c) = &self.config {
                if let Ok(s) = config::Config::builder()
                    .add_source(config::File::new(c.as_str(), config::FileFormat::Yaml))
                    .build()
                {
                        let mut v = s.get::<Vec<VsockParam>>("vms").unwrap();
                        if v.len() == 1 {
                            return v.pop()
                                .map(|vm| VsockConfig::new(vm.guest_cid, vm.socket, vm.uds_path));
                        }
                }
            }
        }

	None
    }

Reply:
@ToolmanP I can see that you have resolved this, but there were few more suggestions in above code block to reduce code size and avoid the extra return None. Won't it be worth doing ?

Reply:
@ToolmanP IIUC @vireshk is referring to:
```
            if let Ok(s) = config::Config::builder()
                .add_source(config::File::new(c.as_str(), config::FileFormat::Yaml))
                .build()
            {
```

Fixed that, I think we are ready to merge this PR :-)